src/main/scala/nodes/util/AllSparseFeatures.scala

```diff
@@ -8,12 +8,20 @@ import scala.reflect.ClassTag
 
 /**
  * An Estimator that chooses all sparse features observed when training,
- * and produces a transformer which builds a sparse vector out of them
+ * and produces a transformer which builds a sparse vector out of them.
+ *
+ * Deterministically orders the feature mappings by earliest appearance in the RDD
  */
 case class AllSparseFeatures[T: ClassTag]() extends Estimator[Seq[(T, Double)], SparseVector[Double]] {
   override def fit(data: RDD[Seq[(T, Double)]]): SparseFeatureVectorizer[T] = {
-    val featureSpace = data.flatMap(_.map(_._1)).distinct()
-        .zipWithIndex().collect().map(x => (x._1, x._2.toInt)).toMap
+    val featureOccurrences = data.flatMap(_.map(_._1))
+    // zip with unique ids and take the smallest unique id for a given feature to get
+    // a deterministic ordering
+    val featuresWithUniqueId = featureOccurrences.zipWithUniqueId().reduceByKey {
+      (x, y) => Math.min(x, y)
```

Comment:
I tried that earlier, but not enough things have an implicit ordering for this to work nicely & easily interpretable w/ our type inference (e.g. Seq, Tuple, etc.). I could replace reduceByKey w/ a localReduceByKey and then the sort would also be local, but more would go to the driver.